crates/oxc_transformer/src/react/refresh.rs

```diff
@@ -0,0 +1,969 @@
+                        Declaration::FunctionDeclaration(func) => {
+                            self.transform_function_declaration(func, ctx)
+                        }
+                        Declaration::VariableDeclaration(variable) => {
+                            self.transform_variable_declaration(variable, ctx)
+                        }
+                        _ => None,
+                    }
+                } else {
+                    None
+                }
+            }
+            Statement::ExportDefaultDeclaration(ref mut stmt_decl) => {
+                match &mut stmt_decl.declaration {
+                    declaration @ match_expression!(ExportDefaultDeclarationKind) => {
+                        let expression = declaration.to_expression_mut();
+                        if !matches!(expression, Expression::CallExpression(_)) {
+                            // For now, we only support possible HOC calls here.
+                            // Named function declarations are handled in FunctionDeclaration.
+                            // Anonymous direct exports like export default function() {}
+                            // are currently ignored.
+                            return None;
+                        }
+
+                        // This code path handles nested cases like:
+                        // export default memo(() => {})
+                        // In those cases it is more plausible people will omit names
+                        // so they're worth handling despite possible false positives.
+                        // More importantly, it handles the named case:
+                        // export default memo(function Named() {})
+                        self.replace_inner_components(
+                            "%default%",
+                            expression,
+                            /* is_variable_declarator */ false,
+                            ctx,
+                        );
+
+                        None
+                    }
+                    ExportDefaultDeclarationKind::FunctionDeclaration(func) => {
+                        if let Some(id) = &func.id {
+                            if !is_componentish_name(&id.name) {
+                                return None;
+                            }
+
+                            return Some(self.create_assignment_expression(id, ctx));
+                        }
+                        None
+                    }
+                    _ => None,
+                }
+            }
+            _ => None,
+        }
+    }
+
+    pub fn transform_statements(
+        &mut self,
+        _stmts: &mut oxc_allocator::Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        self.signature_declarator_items.push(ctx.ast.vec());
+    }
+
+    pub fn transform_statements_on_exit(
+        &mut self,
+        stmts: &mut oxc_allocator::Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
```

Comment:
We have to keep it for now, because we may need to insert new statements under the current statement. Related to #4767.

Both `ExportNamedDeclaration` and `ExportDefaultDeclaration` may confuse you. But we also need to transform the `FunctionDeclaration` inside the `BlockStatement`

For example: 
```ts
function Foo() {
  function name() {}
}
while (true)
  function name() {}
}
```

Reply:
Ah ha I see. Thanks for explaining.